server/src/main/java/org/apache/iotdb/db/mpp/plan/expression/visitor/RemoveWildcardInFilterVisitor.java

```diff
@@ -0,0 +1,148 @@
+    // schemaTree, false);
+    List<Expression> rightExpressions =
+        process(binaryExpression.getRightExpression(), context.notRootClone());
+    //            removeWildcardInFilter(
+    //                    ((BinaryExpression) predicate).getRightExpression(), prefixPaths,
+    // schemaTree, false);
+    if (context.isRoot() && binaryExpression.getExpressionType() == ExpressionType.LOGIC_AND) {
+      List<Expression> resultExpressions = new ArrayList<>(leftExpressions);
+      resultExpressions.addAll(rightExpressions);
+      return resultExpressions;
+    }
+    return reconstructBinaryExpressions(
+        binaryExpression.getExpressionType(), leftExpressions, rightExpressions);
+  }
+
+  @Override
+  public List<Expression> visitFunctionExpression(FunctionExpression predicate, Context context) {
+    List<List<Expression>> extendedExpressions = new ArrayList<>();
+    for (Expression suffixExpression : predicate.getExpressions()) {
+      extendedExpressions.add(
+          process(
+              suffixExpression,
+              new Context(context.getPrefixPaths(), context.getSchemaTree(), false)));
+
+      // We just process first input Expression of AggregationFunction,
+      // keep other input Expressions as origin and bind Type
+      // If AggregationFunction need more than one input series,
+      // we need to reconsider the process of it
+      if (predicate.isBuiltInAggregationFunctionExpression()) {
+        List<Expression> children = predicate.getExpressions();
+        bindTypeForAggregationNonSeriesInputExpressions(
+            predicate.getFunctionName(), children, extendedExpressions);
+        break;
+      }
+    }
+    List<List<Expression>> childExpressionsList = new ArrayList<>();
+    cartesianProduct(extendedExpressions, childExpressionsList, 0, new ArrayList<>());
+    return reconstructFunctionExpressions(predicate, childExpressionsList);
+  }
+
+  @Override
+  public List<Expression> visitTimeSeriesOperand(TimeSeriesOperand predicate, Context context) {
+    PartialPath filterPath = predicate.getPath();
+    List<PartialPath> concatPaths = new ArrayList<>();
+    if (!filterPath.getFirstNode().equals(SqlConstant.ROOT)) {
+      context.getPrefixPaths().forEach(prefix -> concatPaths.add(prefix.concatPath(filterPath)));
+    } else {
+      // do nothing in the case of "where root.d1.s1 > 5"
+      concatPaths.add(filterPath);
+    }
+
+    List<PartialPath> noStarPaths = new ArrayList<>();
+    for (PartialPath concatPath : concatPaths) {
+      List<MeasurementPath> actualPaths =
+          context.getSchemaTree().searchMeasurementPaths(concatPath).left;
+      if (actualPaths.isEmpty()) {
+        return Collections.singletonList(new NullOperand());
+      }
+      noStarPaths.addAll(actualPaths);
+    }
+    return reconstructTimeSeriesOperands(noStarPaths);
+  }
+
+  @Override
+  public List<Expression> visitLeafOperand(LeafOperand leafOperand, Context context) {
```

Comment:
It is confusing to have `visitTimeSeriesOperand()` and `visitLeafOperand()` at the same time. This is handled better in `RemoveWildcardInFilterByDeviceVisitor`, I think.

Reply:
done.